bve-render/src/object.rs

```diff
@@ -0,0 +1,175 @@
+use crate::*;
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct ObjectHandle(pub(crate) u64);
+
+pub struct Object {
+    pub vertex_buffer: Buffer,
+
+    pub index_buffer: Buffer,
+    pub index_count: u32,
+
+    pub texture: u64,
+
+    pub uniform_buffer: Buffer,
+    pub bind_group: BindGroup,
+
+    pub location: Vector3<f32>,
+    pub mesh_center_offset: Vector3<f32>,
+    pub camera_distance: f32,
+
+    pub transparent: bool,
+    pub mesh_transparent: bool,
+}
+
+pub fn is_mesh_transparent(mesh: &[MeshVertex]) -> bool {
+    mesh.iter().any(|v| v.color.w != 0.0 && v.color.w != 1.0)
+}
+
+pub fn convert_mesh_verts_to_render_verts(
+    verts: Vec<MeshVertex>,
+    mut indices: Vec<u32>,
+) -> (Vec<render::Vertex>, Vec<u32>) {
+    // First add the extra faces due to doubling
+    let mut extra_indices = Vec::new();
+
+    for (&i1, &i2, &i3) in indices.iter().tuples() {
+        let v1_double = verts[i1 as usize].double_sided;
+        let v2_double = verts[i2 as usize].double_sided;
+        let v3_double = verts[i3 as usize].double_sided;
+
+        if v1_double || v2_double || v3_double {
+            extra_indices.push(i3);
+            extra_indices.push(i2);
+            extra_indices.push(i1);
+        }
+    }
+
+    // Then convert the verts to the new format
+    let out_verts = verts
+        .into_iter()
+        .map(|v| render::Vertex {
+            pos: v.position.into(),
+            _color: v.color.into(),
+            _normal: v.normal.into(),
+            _tex_coord: v.coord.into(),
+        })
+        .collect_vec();
+
+    indices.extend(extra_indices.into_iter());
+
+    (out_verts, indices)
+}
+
+pub fn find_mesh_center(mesh: &[render::Vertex]) -> Vector3<f32> {
+    let first = if let Some(first) = mesh.first() {
+        *first
+    } else {
+        return Vector3::zero();
+    };
+    // Bounding box time baby!
+    let mut max: Vector3<f32> = first.pos.into();
+    let mut min: Vector3<f32> = first.pos.into();
+
+    for vert in mesh.iter().skip(1) {
+        let pos: Vector3<f32> = vert.pos.into();
+        max = max.zip(pos, |left, right| left.max(right));
+        min = min.zip(pos, |left, right| left.min(right));
+    }
+
+    (max + min) / 2.0
+}
+
+pub fn generate_matrix(mx_view: &Matrix4<f32>, location: Vector3<f32>, aspect_ratio: f32) -> Matrix4<f32> {
+    let mx_projection = cgmath::perspective(cgmath::Deg(55_f32), aspect_ratio, 0.1, 1000.0);
+    let mx_model = Matrix4::from_translation(location);
+    OPENGL_TO_WGPU_MATRIX * mx_projection * mx_view * mx_model
+}
+
+impl Renderer {
+    pub fn add_object(
+        &mut self,
+        location: Vector3<f32>,
+        mesh_verts: Vec<MeshVertex>,
+        indices: &[impl ToPrimitive],
+    ) -> ObjectHandle {
+        self.add_object_texture(location, mesh_verts, indices, &texture::TextureHandle(0))
+    }
+
+    pub fn add_object_texture(
```

Comment:
Looks like you're just setting each object to have one texture?

Might be all that BVE needs, but if you get more complexity - like, if you start having normalmaps - you'll probably want a first-class concept of a material

Reply:
yeah, for now that's fine, but materials are going to be a thing I need going forward.